sway-lib-std/src/storage.sw

```diff
@@ -137,6 +137,118 @@ pub fn clear<T>(key: b256) -> bool {
+///
+/// ### Arguments
+///
+/// * `key` - The storage slot at which the variable will be stored.
+/// * `slice` - The raw_slice to be stored.
+///
+/// ### Examples
+///
+/// ```sway
+/// use std::{alloc::alloc_bytes, storage::{store_slice, get_slice}, constants::ZERO_B256};
+///
+/// let slice = asm(ptr: (alloc_bytes(1), 1)) { ptr: raw_slice };
+/// store_slice(ZERO_B256, slice);
+/// let stored_slice = get_slice(ZERO_B256).unwrap();
+/// assert(slice == stored_slice);
+/// ```
+#[storage(write)]
+pub fn store_slice(key: b256, slice: raw_slice) {
+    // Get the number of storage slots needed based on the size of bytes.
+    let number_of_slots = (slice.len_bytes() + 31) >> 5;
+    let mut ptr = slice.ptr();
+
+    // The capacity needs to be a multiple of 32 bytes so we can 
+    // make the 'quad' storage instruction store without accessing unallocated heap memory.
+    ptr = realloc_bytes(ptr, slice.len_bytes(), number_of_slots * 32);
+
+    // Store `number_of_slots * 32` bytes starting at storage slot `key`.
+    let _ = __state_store_quad(sha256(key), ptr, number_of_slots);
+
+    // Store the length of the bytes
+    store(key, slice.len_bytes());
+}
+
+/// Load a raw_slice from storage.
+///
+/// If no value was previously stored at `key`, `Option::None` is returned. Otherwise,
+/// `Option::Some(value)` is returned, where `value` is the value stored at `key`.
+///
+/// ### Arguments
+///
+/// * `key` - The storage slot to load the value from.
+///
+/// ### Examples
+///
+/// ```sway
+/// use std::{alloc::alloc_bytes, storage::{store_slice, get_slice}, constants::ZERO_B256};
+///
+/// let slice = asm(ptr: (alloc_bytes(1), 1)) { ptr: raw_slice };
+/// store_slice(ZERO_B256, slice);
+/// let stored_slice = get_slice(ZERO_B256).unwrap();
+/// assert(slice == stored_slice);
+/// ```
+#[storage(read)]
+pub fn get_slice(key: b256) -> Option<raw_slice> {
+    // Get the length of the slice that is stored.
+    let len = get::<u64>(key).unwrap_or(0);
+
+    if len > 0 {
+        // Get the number of storage slots needed based on the size.
+        let number_of_slots = (len + 31) >> 5;
+        let ptr = alloc_bytes(number_of_slots * 32);
+
+        // Load the stored slice into the pointer.
+        let _ = __state_load_quad(sha256(key), ptr, number_of_slots);
+
+        Option::Some(asm(ptr: (ptr, len)) { ptr: raw_slice })
+    } else {
+        Option::None
+    }
+}
+
+/// Clear a sequence of bytes storage slots starting at a some key. Returns a Boolean
+/// indicating whether all of the storage slots cleared were previously set.
+///
+/// ### Arguments
+///
+/// * `key` - The key of the first storage slot that will be cleared
+///
+/// ### Examples
+///
+/// ```sway
+/// use std::{alloc::alloc_bytes, storage::{clear_slice, store_slice, get_slice}, constants::ZERO_B256};
+///
+/// let slice = asm(ptr: (alloc_bytes(1), 1)) { ptr: raw_slice };
+/// store_slice(ZERO_B256, slice);
+/// let cleared = clear_slice(ZERO_B256);
```

Comment:
I'd assert that things have been correctly stored as a pre-condition and then clear.